src/random/PrecisionGaussian.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * PrecisionGaussian.cpp
+ *
+ *  Created on: Sept 15, 2014
+ *      Author: yiwu
+ */
+
+#include <cmath>
+#include <cassert>
+#include "PrecisionGaussian.h"
+
+namespace swift {
+namespace random {
+
+PrecisionGaussian::PrecisionGaussian()
+  : dist(0.0,1.0),
+    is_gen_ok(false), is_like_ok(false), is_loglike_ok(false),
+    sqrt2PI(std::sqrt(2*PI)), log2PI(std::log(2*PI)) {
+}
+
+PrecisionGaussian::~PrecisionGaussian() {
+}
+
+void PrecisionGaussian::init(const arma::mat& _mean, const arma::mat& _inv_cov) {
+  mean = _mean;
+  inv_cov = _inv_cov;
+  // Dimension Check
+  if (mean.n_cols > 1)
+    mean = arma::trans(mean);
+  assert(mean.n_cols == 1);
+  assert(mean.n_rows == inv_cov.n_rows && inv_cov.n_cols == inv_cov.n_rows);
+
+  is_gen_ok = false;
+  is_like_ok = true;
+  is_loglike_ok = true;
+
+  // Variables for computing cpd
+  k = mean.n_rows;
+  det_cov = 1 / arma::det(inv_cov);
+  coef = std::pow(sqrt2PI, -k) / std::sqrt(det_cov);
+  log_coef = -k * 0.5 * log2PI - 0.5 * std::log(det_cov);
+}
+
+arma::mat PrecisionGaussian::gen() {
+  if(!is_gen_ok) {
+    c.set_size(mean.n_rows, mean.n_cols);
+    A = arma::flipud(arma::fliplr(arma::inv(arma::chol(inv_cov))));
```

Comment:
From an email I sent out a while back:

Suppose our covariance matrix is Σ, then the Cholesky decomposition gives us Σ = L'L and we generate samples according to L'. With the precision matrix decomposition Σ^(-1) = M'M, and I generate according to L' = M^(-1).

As an example, I've checked the following in armadillo:

![image](https://cloud.githubusercontent.com/assets/1163598/19216615/2a8bd6c6-8d78-11e6-84f9-b15b07f1a9ac.png)


Turns out the matrix is reversed, so I added fliplr and flipud to fix it.

I wrote a quick `test/test_precisiongaussian.blog` comparing the precision gaussian and the original multivar gaussian and the math seems okay. When I didn't include the flipul and fliplr, I had inconsistencies.